rust/projects/parallelism/project.md

```diff
@@ -0,0 +1,447 @@
+
+```rust
+let listener = TcpListener::bind("127.0.0.1:80")?;
+
+for stream in listener.incoming() {
+	let cmd = self.read_cmd(&stream);
+	let resp = self.process_cmd(cmd);
+	self.respond(&stream, resp);
+}
+```
+
+_Well, now you just need to do the same thing, but spawn all the work inside the
+loop into your `NaiveThreadPool`_. The database query and the response are both
+handled on a different thread than the TCP listener.
+
+**Test cases to complete**:
+
+  - TODO @Yilin
+
+Again, you should still have a working client/server key-value store, now
+multithreaded.
+
+
+## Part 4: Creating a real thread pool
+
+So now that you've got your multi-threaded architecture in place, it's time to
+write a real thread pool. You probably wouldn't write your own thread pool in
+practice as there exist thread pool crates that are well-tested, but it
+is a useful exercise to gain experience with parallelism in general. Later in
+this project you will, as we did with the engine in the previous project,
+abstract the thread pool and compare the performance of yours with an existing.
+
+So, what is a thread pool?
+
+It's nothing complicated. Instead of creating a new thread for every
+multi-threaded job to be performed, a thread pool maintains a "pool" of threads,
+and reuses those threads instead of creating a new one.
+
+But why?
+
+It's entirely about performance. It saves a small amount of performance, but
+when writing high-performance applications, every bit counts. Imagine what it
+takes to make a new thread:
+
+You've got to have a call stack for that thread to run on. That's an allocation.
+Allocations are pretty cheap, but not as cheap as not allocation. How that call
+stack is allocated depends on details of the operating system and runtime, but
+can involve locks and syscalls. Syscalls again are not _that_ expensive, but
+they are expensive when we're dealing with Rust levels of performance. Reducing
+syscalls is a common source of easy optimization. That stack then has to be
+carefully initialized so that first [stack _frame_] contains the appropriate
+values for the return pointer and other things. In Rust the stack needs to be
+configured with a [guard page]. That's two more syscalls, [to `mmap` and to
+`mprotect`][mp].
+
+[2mb]: https://github.com/rust-lang/rust/blob/6635fbed4ca8c65822f99e994735bd1877fb063e/src/libstd/sys/unix/thread.rs#L12
+[mp]: https://github.com/rust-lang/rust/blob/6635fbed4ca8c65822f99e994735bd1877fb063e/src/libstd/sys/unix/thread.rs#L315
+
+TODO: illustration?
+
+That's just setting up the callstack. It's at least another syscall to create
+the new thread, at which point the kernel must do its own accounting for the new
+thread.
+
+In Rust libpthread handles most of this complexity.
+
+Then at some point the OS performs a [context switch] onto the new thread. With
+a thread pool, all that setup overhead is only done for a few threads, and
+subsequent jobs are simply context switches into that pool of threads.
```

Comment:
I think the cost of context switch and scheduling among a large number of threads should be mentioned too.

Reply:
Can you say more about what you are thinking? I can imagine explaining that scheduling strategies themselves have a cost that has to be taken into consideration.